File: src/main/java/cz/stasimek/fakturaceeasypeasy/entity/BaseEntity.java
package cz.stasimek.fakturaceeasypeasy.entity;

import cz.stasimek.fakturaceeasypeasy.entity.interfaces.*;
import java.io.Serializable;
import java.time.ZonedDateTime;
import java.util.UUID;
import javax.persistence.Column;
import javax.persistence.GeneratedValue;
import javax.persistence.Id;
import javax.persistence.MappedSuperclass;
import javax.persistence.PrePersist;
import javax.persistence.PreUpdate;
import javax.validation.constraints.PastOrPresent;
import lombok.Getter;
import lombok.Setter;
import org.hibernate.annotations.Filter;
import org.hibernate.annotations.FilterDef;
import org.hibernate.annotations.ParamDef;
import org.hibernate.annotations.SQLDelete;
import org.hibernate.annotations.Where;

@MappedSuperclass
@SQLDelete(sql = "UPDATE #{#entityName} SET deleted = true WHERE id = ?")
@Where(clause = "deleted = false")
@FilterDef(name = "deletedFilter", parameters = @ParamDef(name = "deleted", type = "boolean"))
@Filter(name = "deletedFilter", condition = "deleted = :deleted")
@Getter
@Setter
public abstract class BaseEntity
		implements Identifiable<UUID>, Auditable, SoftDeletable, Serializable {

	@Id
	@Column(columnDefinition = "char(16) for bit data")
	@GeneratedValue
	private UUID id;

	@PastOrPresent
	@Column(nullable = false)
	private ZonedDateTime createdAt;

	@PastOrPresent
	@Column(nullable = false)
	private ZonedDateTime updatedAt;

	@Column
	private boolean deleted = false;

	@Override
	public String toString() {
		return String.format("%s (#%s)", getClass().getSimpleName(), id);
	}

	@PrePersist
	public void prePersist() {
		createdAt = updatedAt = ZonedDateTime.now();
	}

	@PreUpdate
	public void preUpdate() {
		updatedAt = ZonedDateTime.now();
	}

}

File: src/main/kotlin/cz/stasimek/fakturaceeasypeasy/entity/BaseEntity.kt
package cz.stasimek.fakturaceeasypeasy.entity

import cz.stasimek.fakturaceeasypeasy.entity.interfaces.Auditable
import cz.stasimek.fakturaceeasypeasy.entity.interfaces.Identifiable
import cz.stasimek.fakturaceeasypeasy.entity.interfaces.SoftDeletable
import lombok.EqualsAndHashCode
import org.hibernate.annotations.Filter
import org.hibernate.annotations.FilterDef
import org.hibernate.annotations.ParamDef
import org.hibernate.annotations.Where
import java.io.Serializable
import java.time.ZonedDateTime
import java.util.*
import javax.persistence.*
import javax.validation.constraints.PastOrPresent

@MappedSuperclass
// @SQLDelete doesn't work here. Must be in each entity.
// @SQLDelete(sql = "UPDATE #{#entityName} SET deleted = true WHERE id = ?")
@Where(clause = "deleted = false")
@FilterDef(name = "deletedFilter", parameters = [ParamDef(name = "deleted", type = "boolean")])
@Filter(name = "deletedFilter", condition = "deleted = :deleted")
@EqualsAndHashCode(of = ["id"])
abstract class BaseEntity : Identifiable<UUID?>, Auditable, SoftDeletable, Serializable {

    @Id
    @Column(columnDefinition = "char(16) for bit data")
    @GeneratedValue
    open override var id: UUID? = null

    @Column(nullable = false)
    @PastOrPresent
    open override var createdAt: ZonedDateTime? = null

    @Column(nullable = false)
    @PastOrPresent
    open override var updatedAt: ZonedDateTime? = null

    @Column
    open override var deleted = false

    open override fun toString(): String {
        return String.format("%s (#%s)", javaClass.simpleName, id)
    }

    @PrePersist
    private fun prePersist() {
        updatedAt = ZonedDateTime.now()
        createdAt = updatedAt
    }

    @PreUpdate
    private fun preUpdate() {
        updatedAt = ZonedDateTime.now()
    }
}